src/lib/util.ts

```diff
@@ -18,84 +27,59 @@ export function firstMatchingToken<T>(tokenList: string, validValues: T[]): T |
   return parseTokenList(tokenList).find((value) => validValues.includes(value as T)) as T | undefined;
 }
 
-/** Are we able to traverse the DOM? */
-export function isHTMLElement(element: HTMLElement | VirtualElement): boolean {
-  return typeof HTMLElement !== 'undefined' && element instanceof HTMLElement;
-}
-
-/** Normalize HTML Elements */
-export function virtualizeElement(element: HTMLElement | VirtualElement): VirtualElement {
-  // handle HTMLElement or VirtualElement
-  let tagName = '' as TagName;
-  let attributes: VirtualElement['attributes'];
-
-  if (isHTMLElement(element)) {
-    tagName = element.tagName.toLowerCase() as TagName;
-    attributes = {};
-    for (let i = 0; i < (element as HTMLElement).attributes.length; i++) {
-      // biome-ignore lint/style/noNonNullAssertion: This is guaranteed
-      const { name } = (element as HTMLElement).attributes[i]!;
-      attributes[name] = (element as HTMLElement).getAttribute(name);
-    }
-    return { tagName, attributes };
-  }
-
-  if (!element || typeof element !== 'object' || Array.isArray(element) || typeof element.tagName !== 'string') {
-    throw new Error(`Expected { tagName, [attributes] } object, received ${JSON.stringify(element)}`);
+export function getTagName(element: Element | VirtualElement): TagName {
+  if (typeof Element !== 'undefined' && element instanceof Element) {
+    return element.tagName.toLowerCase() as TagName;
   }
+  return element.tagName as TagName;
+}
 
-  return { ...element } as VirtualElement;
+/**
+ * Get attribute from any type of element. Optimized for performance.
+ */
+export function attr(element: Element | VirtualElement, attribute: string) {
+  // Note: this is type-safe; don’t add more runtime checks to satify TypeScript
+  return typeof (element as Element).getAttribute === 'function'
+    ? (element as Element).getAttribute(attribute)
+    : (element as VirtualElement).attributes?.[attribute];
 }
 
 /**
  * Determine accessible names for SOME tags (not all)
  * @see https://www.w3.org/TR/wai-aria-1.3/#namecalculation
  */
-export function calculateAccessibleName(element: VirtualElement): string | undefined {
-  const { tagName, attributes } = element;
+export function calculateAccessibleName(element: Element | VirtualElement, role: RoleData): string | undefined {
+  if (role.nameFrom === 'prohibited') {
+    return;
+  }
+  if (role.nameFrom === 'contents') {
+    return 'innerText' in (element as HTMLElement) ? (element as HTMLElement).innerText : undefined;
+  }
 
+  // for author + authorAndContents, handle special cases first
+  const tagName = getTagName(element);
   switch (tagName) {
-    case 'aside': {
```

Comment:
I was able to clean this up nicely with the newly-added `nameFrom` property in role data, which exists both on `aside` and `section`, as well as others. Improved behavior and reduced code.